docs/blog/posts/sled-integration.md

```diff
@@ -0,0 +1,248 @@
+
+```rust
+/// ProgramInfo stores information about bpf programs that are loaded and managed
+/// by bpfd.
+#[derive(Debug, Serialize, Deserialize, Clone, Default)]
+pub(crate) struct ProgramData {
+    // known at load time, set by user
+    name: String,
+    location: Location,
+    metadata: HashMap<String, String>,
+    global_data: HashMap<String, Vec<u8>>,
+    map_owner_id: Option<u32>,
+
+    // populated after load
+    kernel_info: Option<KernelProgramInfo>,
+    map_pin_path: Option<PathBuf>,
+    maps_used_by: Option<Vec<u32>>,
+
+    // program_bytes is used to temporarily cache the raw program data during
+    // the loading process.  It MUST be cleared following a load so that there
+    // is not a long lived copy of the program data living on the heap.
+    #[serde(skip_serializing, skip_deserializing)]
+    program_bytes: Vec<u8>,
+}
+```
+
+This worked well enough, but as mentioned before the process of flushing the data
+to disk involved manual serialization to JSON, which needed to occur at a specific
+point in time (following program load) which made disaster recovery almost
+impossible and could sometimes result in lost or partially reconstructed state.
+
+With sled the first idea was to completely flatten ALL of bpfman's data into a
+single keyspace, so that `program.name` now simply turns into a `db.get("program_<ID>_name")`,
+however removing all of the core structures would have resulted in a complex diff
+which would have been hard to review and merge.  Therefore a more staged approach
+was taken, the `ProgramData` structure was kept around, and now looks like:
+
+```rust
+/// ProgramInfo stores information about bpf programs that are loaded and managed
+/// by bpfman.
+#[derive(Debug, Clone)]
+pub(crate) struct ProgramData {
+    // Prior to load this will be a temporary Tree with a random ID, following
+    // load it will be replaced with the main program database tree.
+    db_tree: sled::Tree,
+
+    // populated after load, randomly generated prior to load.
+    id: u32,
+
+    // program_bytes is used to temporarily cache the raw program data during
+    // the loading process.  It MUST be cleared following a load so that there
+    // is not a long lived copy of the program data living on the heap.
+    program_bytes: Vec<u8>,
+}
+```
+
+All of the fields are now removed in favor of a private reference to the unique
+[`sled::Tree`](https://docs.rs/sled/latest/sled/struct.Tree.html) instance for
+this `ProgramData` which is named using the unique kernel id for the program.
+Each `sled::Tree` represents a single logical keyspace / namespace / bucket which
+allows key generation to be kept simple, i.e `db.get("program_<ID>_name")`
+now can be `db_tree_prog_0000.get("program_name)`. Additionally getters and setters are now
+built for each existing field so that access to the db can be controlled
+and the existing degree of type safety can be maintained:
```

Comment:
It's more to abstract the serialization/deserialization from the caller of the getter/setter.